util.go

```diff
@@ -32,122 +32,152 @@ func generateChallengeKey() (string, error) {
 }
 
 // Octet types from RFC 2616.
-var octetTypes [256]byte
+//
+// OCTET      = <any 8-bit sequence of data>
+// CHAR       = <any US-ASCII character (octets 0 - 127)>
+// CTL        = <any US-ASCII control character (octets 0 - 31) and DEL (127)>
+// CR         = <US-ASCII CR, carriage return (13)>
+// LF         = <US-ASCII LF, linefeed (10)>
+// SP         = <US-ASCII SP, space (32)>
+// HT         = <US-ASCII HT, horizontal-tab (9)>
+// <">        = <US-ASCII double-quote mark (34)>
+// CRLF       = CR LF
+// LWS        = [CRLF] 1*( SP | HT )
+// TEXT       = <any OCTET except CTLs, but including LWS>
+// separators = "(" | ")" | "<" | ">" | "@" | "," | ";" | ":" | "\" | <">
+//              | "/" | "[" | "]" | "?" | "=" | "{" | "}" | SP | HT
+// token      = 1*<any CHAR except CTLs or separators>
+// qdtext     = <any TEXT except <">>
 
-const (
-	isTokenOctet = 1 << iota
-	isSpaceOctet
-)
-
-func init() {
-	// From RFC 2616
-	//
-	// OCTET      = <any 8-bit sequence of data>
-	// CHAR       = <any US-ASCII character (octets 0 - 127)>
-	// CTL        = <any US-ASCII control character (octets 0 - 31) and DEL (127)>
-	// CR         = <US-ASCII CR, carriage return (13)>
-	// LF         = <US-ASCII LF, linefeed (10)>
-	// SP         = <US-ASCII SP, space (32)>
-	// HT         = <US-ASCII HT, horizontal-tab (9)>
-	// <">        = <US-ASCII double-quote mark (34)>
-	// CRLF       = CR LF
-	// LWS        = [CRLF] 1*( SP | HT )
-	// TEXT       = <any OCTET except CTLs, but including LWS>
-	// separators = "(" | ")" | "<" | ">" | "@" | "," | ";" | ":" | "\" | <">
-	//              | "/" | "[" | "]" | "?" | "=" | "{" | "}" | SP | HT
-	// token      = 1*<any CHAR except CTLs or separators>
-	// qdtext     = <any TEXT except <">>
-
-	for c := 0; c < 256; c++ {
-		var t byte
-		isCtl := c <= 31 || c == 127
-		isChar := 0 <= c && c <= 127
-		isSeparator := strings.IndexRune(" \t\"(),/:;<=>?@[]\\{}", rune(c)) >= 0
-		if strings.IndexRune(" \t\r\n", rune(c)) >= 0 {
-			t |= isSpaceOctet
-		}
-		if isChar && !isCtl && !isSeparator {
-			t |= isTokenOctet
-		}
-		octetTypes[c] = t
-	}
-}
-
-func skipSpace(s string) (rest string) {
-	i := 0
-	for ; i < len(s); i++ {
-		if octetTypes[s[i]]&isSpaceOctet == 0 {
-			break
+func skipSpace(s string) string {
+	for i := 0; i < len(s); i++ {
+		switch s[i] {
+		case ' ', '\t', '\r', '\n':
+		default:
+			return s[i:]
 		}
 	}
-	return s[i:]
+	return ""
 }
 
 func nextToken(s string) (token, rest string) {
 	i := 0
+loop:
 	for ; i < len(s); i++ {
-		if octetTypes[s[i]]&isTokenOctet == 0 {
+		c := s[i]
+		if c <= 31 || c >= 127 { // control characters & non-ascii are not token octets
 			break
 		}
+		switch c { //separators are not token octets
```

Comment:
Can you add a test-case into the nextTokenTests for the separator parsing case?

Reply:
There are 2 test cases for the separator parsing case. there are no tests for the control characters and non-ascii, because I am not sure what is appropriate

Reply:
Does the calling code remove the separator?

```
{"other,websocket,more", "other", ",websocket,more"},
{"websocket,more", "websocket", ",more"},
{"more", "more", ""},
```

The "next" token has a leading "," but the input of the next test case has that stripped.